os/fs/smartfs/smartfs_utils.c

```diff
@@ -1268,6 +1268,206 @@ int smartfs_createentry(struct smartfs_mountpt_s *fs, uint16_t parentdirsector,
 	return ret;
 }
 
+#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
+/****************************************************************************
+ * Name: smartfs_findvalidentry
+ * 
+ * Description: Finds an invalid entry in the parent directory that can be
+ *              used to create a new file entry without allocating a new
+ *              sector.
+ *
+ ****************************************************************************/
+
+bool smartfs_findinvalidentry(struct smartfs_mountpt_s *fs, uint16_t parentdirsector, struct smartfs_entry_s *invalidentry)
+{
+	uint16_t ret;
+	struct smart_read_write_s readwrite;
+	uint16_t psector;
+	uint16_t nextsector;
+	uint16_t offset;
+	uint16_t entrysize;
+	struct smartfs_entry_header_s *entry;
+	struct smartfs_chain_header_s *chainheader;
+
+	psector = parentdirsector;
+	entrysize = sizeof(struct smartfs_entry_header_s) + fs->fs_llformat.namesize;	
+
+	while (psector != SMARTFS_ERASEDSTATE_16BIT) {
+		/* Read the next sector */
+
+		readwrite.logsector = psector;
+		readwrite.count = fs->fs_llformat.availbytes;
+		readwrite.offset = 0;
+		readwrite.buffer = (uint8_t *)fs->fs_rwbuffer;
+		ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
+		if (ret < 0) {
+			return FALSE;
```

Comment:
This is file operation,  VFS will save errno. please see function open in fs_open.
You can change it easily by refer to smartfs_finddirentry.

Reply:
Agreed